bin/build_cmd.ml

```diff
@@ -82,12 +82,40 @@ let run_build_system ~common ~request =
       Fiber.return ())
 ;;
 
-let run_build_command_poll_eager ~(common : Common.t) ~config ~request : unit =
+let lock_ocamlformat () =
+  if Lazy.force Lock_dev_tool.is_enabled
+  then
+    (* Note that generating the ocamlformat lockdir here means
+             that it will be created when a user runs `dune fmt` but not
+             when a user runs `dune build @fmt`. It's important that
+             this logic remain outside of `dune build`, as `dune
+             build` is intended to only build targets, and generating
+             a lockdir is not building a target. *)
+    Lock_dev_tool.lock_ocamlformat () |> Memo.run
+  else Fiber.return ()
+;;
+
+let run_build_command_poll_eager
+  ~with_lock_ocamlformat
```

Comment:
I'm not a fan of introducing special behaviour into all these functions when a flag is passed. I think it muddies what was once a fairly clean interface with a special case which makes it harder to conceptualize what these functions do.

Reply:
Agreed. A flag is certainly better than a hook, but we should try to do better.

Reply:
I think we could at least agree upon https://github.com/ocaml/dune/pull/10951/commits/69ea9cf041bcea3e2d3ec1d2de5ebebd3adb7aaa at least or start from that.

Reply:
What was the issue with running the scheduler multiple times? It seems like that would also solve this problem and IMO would be preferable to either a hook or a flag.

Reply:
I'm also OK to run the scheduler once because it avoid loading the project twice than having a refactoring preference.

Reply:
If we're not going to solve this with a hook, a flag, or with running the scheduler twice, then I'm not sure what other options are possible. If/when we make solving dependencies into a build rule/action then this will become simpler. In the meantime it sounds like a flag is the least bad option. @rgrinberg how do you feel about merging this with the flag for now?

Reply:
I looked at the code again and I don't think this is adequate. Here's a plan that addresses most of my concerns:

1. Remove watch mode support from `fmt`. There's no point to it as it's essentially racing against the user to edit source files
2. Implement `dune fmt` by running the scheduler and then running the lock + build inside.

Reply:
> There's no point to it as it's essentially racing against the user to edit source files

Is it because no one using `fmt` on watch mode, this is not addressed before ?

Reply:
Yes, probably nobody is using fmt in watch mode. It would be rather frustrating as it's always racing to update the file as you're editing it, possibly overwriting your changes.